sample/rules.go

```diff
@@ -190,11 +201,30 @@ func ruleMatchesSpanInTrace(trace *types.Trace, rule *config.RulesBasedSamplerRu
 		return true
 	}
 
+	// if a condition contains a root prefix, store index after first check so that we only check it once
+	var seen []int
+
 	for _, span := range trace.GetSpans() {
 		ruleMatched := true
-		for _, condition := range rule.Conditions {
-			// whether this condition is matched by this span.
-			value, exists := extractValueFromSpan(trace, span, condition, checkNestedFields)
+		for i, condition := range rule.Conditions {
+			var value any
+			var exists bool
+			if slices.Contains(seen, i) {
```

Comment:
Ok, now that I understand what "seen" is, we still have a problem. What it's trying to do is not yet correct.

If either all or none of the entries in Fields are in `root`, I think it works. But imagine that your condition is:
```
Fields: http.status, root.http.status
Operator: '='
Value: 500
```
You look at the first span, and neither root nor the span has a status code of 500, and you mark it as "seen". The next span's code IS 500, but because `seen` has been set for this rule, you won't detect it.

This is a problem in both the Trace and Span scopes. 

I know this is frustrating that we keep iterating on this, but I didn't understand until now what you were trying to do. 

I think what we should do is:
* Write one or more tests that would detect the above scenario.
* `extractValueFromSpan` needs to be fixed -- `checkedRootPrefix` needs to be renamed to `usedRootPrefix` and should only be true on return if `value` came from a field using `root.` rather than the span being tested.
* But now that we've done that, we're not going to use it yet. We should call it without optimizations in both scopes and prove that everything works.
* Then, and only then, should we consider whether there are paths to avoid evaluating the root on every span. I think it will be easier to think about once we have the basic logic correct.

I apologize for any frustration but we have to get this right.